### PR TITLE
Fix hang when waiting for replication threads.

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
### Details
This change is not big, but it's super nuanced. Let me try and explain as much as I can. From the top down in the diff:

1. We undo this change: https://github.com/Expensify/Bedrock/pull/1265

Why do that? Because in researching how this works, I came across this change: https://github.com/Expensify/Bedrock/pull/919

That's where we added the above behavior, and I think we actually want to keep it. Read the issue there to see what it does, but basically I think the change in #1265 was the wrong way to do that. More on this below.

2. The second block just adds some logging to a while loop so we can see if we get stuck in it. This seems to be what's happening in issue 209479 that this PR fixes, so it'll be nice to know if we actually get stuck there.

3. Ok, this is the second part of undoing 1265, and handling it better. Now, if we get to where we'd approve a quorum transaction but we've lost the connection to leader, we just warn and continue along. The case that we're trying to handle is that we are late approving a quorum transaction, and leader has already sent us a bunch more commits (because quorum was handled by other nodes). In this case, we want to commit this transaction, even if leader has since disconnected. So long as leader sent us the COMMIT message for this transaction, it will still get committed. This unblocks all other replication threads, so that the change above, where we only cancel commits that leader hasn't said are complete can complete.

This leaves the possibility that leader sends us a quorum transaction, we approve, and we get disconnected before leader sends us the COMMIT for this transaction, in which case, we should still fall out of FOLLOWING and cancel all outstanding commit threads, including this one that's still waiting on this commit from leader.

4. This gets to the meat of fixing the actual linked issue. It seems that sometimes replication threads can get stuck. This adds a timeout so every second they re-check their conditions, and thus shouldn't be "stuck" for more than a second. It doesn't actually address why they get stuck in the first place, which I'm not sure about. It warns if we hit  a timeout in suspicious circumstances, which is where we likely would have gotten stuck before.

5. Setting `_globalResult` to be `atomic` *might* fix the above issue of getting stuck. I'm not actually sure. It seems possible that this value could get cached in the CPU, and then when it was updated by one thread, the other thread that was waiting for it would be interrupted (where we added the check for the timeout above). Whenthis thread was interrupted, it might hit the top of the loop, and check for `_globalResult == RESULT::CANCELED` by looking at a value of `_globalResult` that was in the CPU cache, not yet updated, even though the other thread set this value before interrupting this thread.

If that were to happen, this thread would jump back into it's `wait()` but, being that the cancel had already happened, it could just sit there forever.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/209479

### Tests
Only existing tests.

### CODE REVIEW

File: sqlitecluster/SQLiteNode.cpp [hunk 1]
```diff
@@ -1980,15 +1980,20 @@ void SQLiteNode::_changeState(SQLiteNode::State newState) {
...
             _localCommitNotifier.cancel(cancelAfter);
             _leaderCommitNotifier.cancel(cancelAfter);
 
             // Polling wait for threads to quit. This could use a notification model such as with a condition_variable,
             // which would probably be "better" but introduces yet more state variables for a state that we're rarely
             // in, and so I've left it out for the time being.
+            size_t infoCount = 1;
             while (_replicationThreadCount) {
+                if (!(infoCount % 100)) {
+                    SINFO("Waiting for " << _replicationThreadCount << "remaining replication threads.");
```
Comment (rafecolton) [inline sqlitecluster/SQLiteNode.cpp:1994]:
```suggestion
                    SINFO("Waiting for " << _replicationThreadCount << " remaining replication threads.");
```

File: sqlitecluster/SQLiteNode.cpp [hunk 2]
```diff
@@ -1980,15 +1980,20 @@ void SQLiteNode::_changeState(SQLiteNode::State newState) {
...
+            SINFO("Replication threads should exit, canceling commits after current leader commit " << cancelAfter);
             _localCommitNotifier.cancel(cancelAfter);
             _leaderCommitNotifier.cancel(cancelAfter);
 
             // Polling wait for threads to quit. This could use a notification model such as with a condition_variable,
             // which would probably be "better" but introduces yet more state variables for a state that we're rarely
             // in, and so I've left it out for the time being.
+            size_t infoCount = 1;
             while (_replicationThreadCount) {
+                if (!(infoCount % 100)) {
```
Comment (rafecolton) [inline sqlitecluster/SQLiteNode.cpp:1993]: NAB I think `if (infoCount % 100 == 0)` would be more intuitive

File: sqlitecluster/SQLiteSequentialNotifier.cpp [hunk 3]
```diff
@@ -37,7 +37,19 @@ SQLiteSequentialNotifier::RESULT SQLiteSequentialNotifier::waitFor(uint64_t valu
...
+        cv_status result = state->waitingThreadConditionVariable.wait_for(lock, 1s);
+        if (result == cv_status::timeout) {
+            // So, normally, we should only get woken up if something has happened. If we get woken up because of a
+            // timeout, that's fundamentally fine, we could just still be waiting for that thing to happen.
+            // But if one of the things we're tracking has changed, and we got woken up from a timeout, not from that
+            // change, that's worrisome, and indicates that maybe there's a condition in which we can get stuck. Note
+            // that this isn't 100%, we could get woken up here by chance and one of these conditions could change
+            // immediately following that wakeup, so there's a small but nonzero chance of this log line firing in a
+            // valid case.
+            if (_globalResult == RESULT::CANCELED || state->result != RESULT::UNKNOWN) {
```
Comment (rafecolton) [inline sqlitecluster/SQLiteSequentialNotifier.cpp:49]: Do you need to check `|| _globalResult == RESULT::UNKNOWN` here too?

File: sqlitecluster/SQLiteSequentialNotifier.h [hunk 4]
```diff
@@ -53,7 +53,7 @@ class SQLiteSequentialNotifier {
     uint64_t _value;
 
     // If there is a global result for all pending operations (i.e., they've been canceled), that is stored here.
-    RESULT _globalResult;
+    atomic<RESULT> _globalResult;
```
Comment (rafecolton) [inline sqlitecluster/SQLiteSequentialNotifier.h:56]: If you change this and the issue appears to go away, we won't know if it's actually fixed or not. I'd love to do blue/green with just the log line and then add this after, but I'm not sure how easy it is to reproduce. What do you think?

File: sqlitecluster/SQLiteSequentialNotifier.cpp (hunk 3 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteSequentialNotifier.cpp:49]: It's only ever CANCELED or UNKNOWN

**Review by rafecolton — APPROVED** [review]

**Pull request merged**